framework/src/media/InputHandler.cpp

```diff
@@ -71,12 +72,118 @@ bool InputHandler::doStandBy()
 	return true;
 }
 
+bool InputHandler::open()
+{
+	if (!mInputDataSource) {
+		meddbg("mInputDataSource is nullptr!\n");
+		return false;
+	}
+
+	if (!getStreamBuffer()) {
+		auto streamBuffer = StreamBuffer::Builder()
+							.setBufferSize(CONFIG_HANDLER_STREAM_BUFFER_SIZE)
+							.setThreshold(CONFIG_HANDLER_STREAM_BUFFER_THRESHOLD)
+							.build();
+		if (!streamBuffer) {
+			meddbg("streamBuffer is nullptr!\n");
+			return false;
+		}
+		setStreamBuffer(streamBuffer);
+	}
+
+	// Open data soruce
+	if (!(mInputDataSource->isPrepared() || mInputDataSource->open())) {
+		meddbg("open data source failed!\n");
+		return false;
+	}
+
+	// Identify audio type and other informations if unspecified
+	audio_type_t audioType = mInputDataSource->getAudioType();
+	unsigned int channels = mInputDataSource->getChannels();
+	unsigned int sampleRate = mInputDataSource->getSampleRate();
+	if (audioType == AUDIO_TYPE_UNKNOWN || channels == 0 || sampleRate == 0) {
+		// Preload data from data source
+		ssize_t buflen = CONFIG_DATASOURCE_PREPARSE_BUFFER_SIZE;
+		unsigned char *bufptr = new unsigned char[buflen];
+		if (!bufptr) {
+			meddbg("Out of memory\n");
+			return false;
+		}
+		buflen = mInputDataSource->read(bufptr, buflen);
+		if (buflen <= 0) {
+			meddbg("Read data source failed\n");
+			delete[] bufptr;
+			return false;
+		}
+		// Identify audio type from stream if it's not specified
+		if (audioType == AUDIO_TYPE_UNKNOWN) {
+			audioType = utils::getAudioTypeFromStream(bufptr, (size_t)buflen);
+			if (audioType == AUDIO_TYPE_UNKNOWN) {
+				meddbg("Get audio type from stream failed\n");
+				delete[] bufptr;
+				return false;
+			}
+		}
+		// Always perform header parsing (App can not specify only channels and sample rate, but audio type.)
+		audio_format_type_t pcmFormat = mInputDataSource->getPcmFormat();
+		if (!utils::buffer_header_parsing(bufptr, buflen, audioType, &channels, &sampleRate, &pcmFormat)) {
+			meddbg("Audio header parsing failed\n");
+			delete[] bufptr;
+			return false;
+		}
+		// Update data source attributes
+		mInputDataSource->setAudioType(audioType);
+		mInputDataSource->setChannels(channels);
+		mInputDataSource->setSampleRate(sampleRate);
+		mInputDataSource->setPcmFormat(pcmFormat);
+		medvdbg("channels %u sampleRate %u pcmFormat %d\n", channels, sampleRate, pcmFormat);
+		// Write data to preloaded-buffer for later use
+		mPreloadBuffer = StreamBuffer::Builder()
+						.setBufferSize(buflen)
+						.setThreshold(buflen)
+						.build();
+		if (!mPreloadBuffer) {
+			meddbg("Out of memory\n");
+			delete[] bufptr;
+			return false;
+		}
+		mPreloadBuffer->write(bufptr, buflen);
+		delete[] bufptr;
+	}
+
+	// Register decoder
+	if (!registerCodec(audioType, channels, sampleRate)) {
```

Comment:
StreamHandler doing this, if you moved to this place, then another must be removed.

Reply:
I removed this logic from InputHandler, old logic in StreamHandler.